packages/expo-updates/ios/EXUpdates/AppLoader/EXUpdatesAppLoader.m

```diff
@@ -0,0 +1,241 @@
+
+@property (nonatomic, strong) NSLock *arrayLock;
+
+@end
+
+static NSString * const kEXUpdatesAppLoaderErrorDomain = @"EXUpdatesAppLoader";
+
+@implementation EXUpdatesAppLoader
+
+- (instancetype)init
+{
+  if (self = [super init]) {
+    _assetsToLoad = [NSMutableArray new];
+    _erroredAssets = [NSMutableArray new];
+    _finishedAssets = [NSMutableArray new];
+    _existingAssets = [NSMutableArray new];
+    _arrayLock = [[NSLock alloc] init];
+  }
+  return self;
+}
+
+- (void)_reset
+{
+  _assetsToLoad = [NSMutableArray new];
+  _erroredAssets = [NSMutableArray new];
+  _finishedAssets = [NSMutableArray new];
+  _existingAssets = [NSMutableArray new];
+  _updateManifest = nil;
+  _successBlock = nil;
+  _errorBlock = nil;
+}
+
+# pragma mark - subclass methods
+
+- (void)loadUpdateFromUrl:(NSURL *)url
+                  success:(EXUpdatesAppLoaderSuccessBlock)success
+                    error:(EXUpdatesAppLoaderErrorBlock)error
+{
+  @throw [NSException exceptionWithName:NSInternalInconsistencyException reason:@"Should not call EXUpdatesAppLoader#loadUpdate -- use a subclass instead" userInfo:nil];
+}
+
+- (void)downloadAsset:(EXUpdatesAsset *)asset
+{
+  @throw [NSException exceptionWithName:NSInternalInconsistencyException reason:@"Should not call EXUpdatesAppLoader#loadUpdate -- use a subclass instead" userInfo:nil];
+}
+
+# pragma mark - loading and database logic
+
+- (void)startLoadingFromManifest:(EXUpdatesUpdate *)updateManifest
+{
+  if (![self _shouldStartLoadingUpdate:updateManifest]) {
+    if (_successBlock) {
+      _successBlock(nil);
+    }
+    return;
+  }
+
+  EXUpdatesDatabase *database = [EXUpdatesAppController sharedInstance].database;
+  dispatch_async(database.databaseQueue, ^{
+    NSError *existingUpdateError;
+    EXUpdatesUpdate *existingUpdate = [database updateWithId:updateManifest.updateId error:&existingUpdateError];
+
+    if (existingUpdate && existingUpdate.status == EXUpdatesUpdateStatusReady) {
+      if (self->_successBlock) {
+        dispatch_async(dispatch_get_global_queue(DISPATCH_QUEUE_PRIORITY_DEFAULT, 0), ^{
+          self->_successBlock(updateManifest);
+        });
+      }
+      return;
+    }
+
+    if (existingUpdate) {
+      // we've already partially downloaded the update.
+      // however, it's not ready, so we should try to download all the assets again.
+      self->_updateManifest = updateManifest;
+    } else {
+      if (existingUpdateError) {
+        NSLog(@"Failed to select old update from DB: %@", existingUpdateError.localizedDescription);
+      }
+      // no update already exists with this ID, so we need to insert it and download everything.
+      self->_updateManifest = updateManifest;
+      NSError *updateError;
+      [database addUpdate:self->_updateManifest error:&updateError];
+
+      if (updateError) {
+        dispatch_async(dispatch_get_global_queue(DISPATCH_QUEUE_PRIORITY_DEFAULT, 0), ^{
```

Comment:
Might want to do this dispatch inside _finishWithError when calling the block, though it's probably not a big deal either way.